libs/lib-client/diagram/src/lib/diagram/diagram.component.ts

```diff
@@ -62,12 +68,21 @@ export class DiagramComponent implements OnChanges, OnDestroy {
   private updateMatrix() {
     const tileMatrix = makeRenderableMatrix(this.flow, 10, this.isReadOnly);
     this.rowIndexService.updateRowIndexes(tileMatrix);
-    if (tileMatrix.length > 0) {
+    if (tileMatrix.length > 1) {
+      const rowParentsList = this.updateRowParentsList(tileMatrix);
       // matrix is reversed to make sure html stack order always goes from bottom to top
       // i.e. top rows are rendered in front of bottom rows, this ensures branches don't display on top of the tiles above
       this.tileMatrix = tileMatrix.reverse();
-    } else if (!this.isReadOnly) {
+      this.rowParentsList = rowParentsList.reverse();
+    } else if (tileMatrix.length === 0 && !this.isReadOnly) {
       this.tileMatrix = EMPTY_MATRIX;
+    } else {
+      this.tileMatrix = tileMatrix;
     }
   }
+
+  private updateRowParentsList(tileMatrix) {
+    const rowIndexes = this.rowIndexService.getRowIndexes();
+    return this.dragService.getAllRowsParentTiles(tileMatrix, rowIndexes);
```

Comment:
I am bit worried that the `getAllRowsParentTiles` is taking in `tileMatrix`. It creates additional burden on other consumers of this service to create a similar tileMatrix structure for reusing other features of drag-tile service. It is better if the service takes `GraphNode` as it is common structure for all diagrams. May be we need to revisit it later. Create that parent-child rows mapping calculation based on `GraphNode / GraphNode.nodes` data 
```suggestion
    return this.dragService.getAllRowsParentTiles(tileMatrix, rowIndexes);
```
```suggestion
    return this.dragService.getAllRowsParentTiles(this.flow.nodes, getRowIndexOf);
```

Reply:
Okay, i think it will be a big change to implement. I will update the ticket to implement logic this way as future enhancement.

Reply:
Resolving this conversation